### DISCUSSION TITLE
Embed scripting/tooling with inputs into Go

### DISCUSSION THREAD

**Discussion by jlarfors** [cue-lang/cue#2072] [Q&A]
Hello!

## Background

We are prototyping a self-service tool that allows users to define Request Templates using Cue definitions.
The Go backend generates a OpenAPI specs for each of these, and our JS frontend renders a form for end users to fill in and make requests.
Next step is to have an executor/worker that actually does something once a request is approved, and Cue's scripting/tooling layer looks very appealing.

So what we are trying to achieve: ability to invoke Cue's scripting interface and input values (objects).

### Using Cue scripting

After digging through the Cue codebase I found that the only way to invoke Cue's scripting interface from a Go program was to call it at the higest level, e.g.

```go
package cuework

import (
	"context"
	"testing"

	"cuelang.org/go/cmd/cue/cmd"
	"github.com/stretchr/testify/require"
)

func TestCueScript(t *testing.T) {
	cmd, err := cmd.New([]string{"cmd", "autodep"})
	require.NoError(t, err)
	runErr := cmd.Run(context.TODO())
	require.NoError(t, runErr)
}
```

This is kind of nice, but then comes the struggle of providing input. Using tagged fields and the `-t` flag does not work for objects (at least not nicely, if at all possible).
Ideas I had were to dump a JSON file that can be read, or dump some Cue file containing the value that can then be used.
Neither are ideal.

### Using Cue flow

Another option I found was to use the [flow package](https://pkg.go.dev/cuelang.org/go/tools/flow).
However, I found that I cannot use the `tool` package (i.e. [file](https://pkg.go.dev/cuelang.org/go/pkg/tool/file)  or [http](https://pkg.go.dev/cuelang.org/go@v0.4.3/pkg/tool/http)) because the types are private, and the [task package](https://pkg.go.dev/cuelang.org/go@v0.4.3/internal/task) is internal.
And then it started to feel like writing a lot more than I was hoping for. Is this correct, or did I miss something obvious that means calling REST APIs, reading files, etc can be very easily added?

With this approach it's very easy to fill Cue values and generally make the end user experience nicer, which is good.

## Feedback

Maybe someone who knows more about this can help. I am happy to provide more information about what we are trying to achieve if it's not clear.
I think it would be nice if we do not create a new syntax and can use Cue's native scripting syntax (e.g. `commad: mycmd: { ... }`). However it would be fine to create something ourselves using the `flow` package if it didn't require writing something too extensive (my coworkers joked that we'd be writing our own dagger.io tool, and it started to feel like it...)

Thanks in advance for anyone who tries to help!

Jacob

**Comment by jlarfors** [comment]
After some more research+hacking, I realised that the [flow package](https://pkg.go.dev/cuelang.org/go/tools/flow) is absolutely the way for me to proceed.

Main reasons for this are:
1. Easy to fill inputs with a [flow.RunnerFunc](https://pkg.go.dev/cuelang.org/go/tools/flow#RunnerFunc) (you just have to identify the inputs to fill)
2. Easy to get outputs: `result := controller.Value()`
3. It's very low effort to use the `flow` package to create some sort of DSL or custom task script. I actually find it very pleasant to work with (well done!).

The only downside is not being able to use the [tool package](https://pkg.go.dev/cuelang.org/go@v0.4.3/pkg/tool) together with the `flow` package.
If it's not possible today, and there's nothing against adding it, I'd happily contribute to make it happen!

**Reply by myitcv** [reply]
> After some more research+hacking, I realised that the [flow package](https://pkg.go.dev/cuelang.org/go/tools/flow) is absolutely the way for me to proceed.

Yes, this is exactly what you want.

> The only downside is not being able to use the [tool package](https://pkg.go.dev/cuelang.org/go@v0.4.3/pkg/tool) together with the `flow` package.

It's absolutely possible if your runner uses the same logic for finding tasks as `cmd/cue cue`. You can either use exactly the same logic, or you can have different logic and provide a `TaskFunc` that finds tasks based on a hybrid approach. Hence this is something that's entirely under your control as the author of the program using `tools/flow`. 

What we will be doing as part of https://github.com/cue-lang/cue/issues/1325 is making the declaration of `tool` packages more principled. This would also make it easier to expose those packages for reuse. 

Alternatively you can request that packages be added to `cuelang.org/go/pkg/tool` - it's by no means complete!

**Reply by jlarfors** [reply]
Thanks for the reply. Maybe you can actually tell me how I can use the `tool` packages then :)

Here's a minimal example, based on the example on the `flow` package page, and digging through the `cmd/custom.go` file in the cue repository.

How do I actually "run" or "execute" the builtin tasks in `taskFunc`? Would appreciate the help, probably I am missing something obvious...

```cue
// File: testdata/example.cue
import "tool/cli"

a: {
	do: cli.Print & {
		text: "gimme the loot!"
	}
}
```

```go
// File: flow_test.go
package main

import (
	"context"
	"fmt"
	"log"
	"testing"

	"cuelang.org/go/cue"
	"cuelang.org/go/cue/cuecontext"
	"cuelang.org/go/cue/load"
	"cuelang.org/go/tools/flow"
	"github.com/stretchr/testify/require"
)

func TestFlow(t *testing.T) {
	cueCtx := cuecontext.New()
	insts := load.Instances([]string{"./testdata/example.cue"}, nil)
	require.Len(t, insts, 1)
	val := cueCtx.BuildInstance(insts[0])
	require.NoError(t, val.Err())
	controller := flow.New(nil, val, taskFunc)
	if err := controller.Run(context.Background()); err != nil {
		log.Fatal(err)
	}
}

func taskFunc(v cue.Value) (flow.Runner, error) {

	if !isTask(v) {
		return nil, nil
	}
	return flow.RunnerFunc(func(t *flow.Task) error {
		fmt.Println("Task at ", t.Value().Path().String())
		// How do I actually make this task "do" something??
		return nil
	}), nil
}

func isTask(v cue.Value) bool {
	if v.Kind() != cue.StructKind {
		return false
	}
	if v.LookupPath(cue.ParsePath("$id")).Exists() {
		return true
	}
	return false
}

```

**Reply by myitcv** [reply]
> Maybe you can actually tell me how I can use the `tool` packages then :)

Hmm, I might be guilty here of suggesting something without actually trying it first!

It seems that the relevant bits you need are actually `internal` at the moment, so not able to be reused right now.

Stepping back a bit here: are you looking to define your own `tool`-like tasks, or simply reuse those in the existing `tool/*` packages?

I'd like to make sure we fully understand what you're after here, to ensure we have open issues for the relevant missing bits.

> Using tagged fields and the `-t` flag does not work for objects (at least not nicely, if at all possible).

This is possible using `string` type fields with `json.Unmarshal`:

```
exec cue export -t 'x={"y":5}' -e y
cmp stdout stdout.golden

-- x.cue --
package x

import "encoding/json"

x: string @tag(x)
y: json.Unmarshal(x).y
-- stdout.golden --
5
```

**Reply by jlarfors** [reply]
OK, no worries and that was my discovery as well (everything relevant is internal or private).

Here's an example cue file for the current prototype:

```cue
package basic

import "tool/http"

#ExampleRequest: {
	kind: "ExampleRequest"
	spec: {
		name: =~"^[A-Za-z0-9-]+$"
	}
}

task: example: {
	input: #ExampleRequest

	step: restAPI: http.Get & {
		url: "<url>"
	}

	output: {
		status: step.restAPI.response.status
	}
}
```

The definition `#ExampleRequest` defines the "template" for a request that can be made by end users. End users get a nice form in the web frontend to fill in the `spec` (in this case, just one string field `name`).
There is then an approval flow (manual or automatic) and when the request is approved, a trigger is sent to a worker who runs the `task.example` (because the `input` matches the incoming request).
The `task.example.output` is returned and stored in the backend together with the approved request.

We have implemented all this already and it's working  using the `flow` package: filling the `input.spec` with the user's submitted values and getting the `output` from the end of the run. And this is where the `flow` package seems really nice!

So the only part missing now is enabling the `flow` engine to do things like: HTTP requests, execute CLIs, etc. I am expecting to have to write custom tasks also, but it would be great to reuse the builtin tasks and also contribute back to those with extra things we might need. I am so certain this will help others in the future.

So from my very newbie CUE experience, my request would be: how to expose the `tool` packages so that we can use them in custom `flow` packages.

Let me know what you think about this, and any comments/feedback/whatever is welcome as there isn't much material on CUE online to compare this with :)
